lib/rules/utils/ast-utils.js

```diff
@@ -1015,6 +1015,114 @@ function isDirective(node) {
     return node.type === "ExpressionStatement" && typeof node.directive === "string";
 }
 
+/**
+ * Tests if a node appears at the beginning of an ancestor ExpressionStatement node.
+ * @param {ASTNode} node The node to check.
+ * @returns {boolean} Whether the node appears at the beginning of an ancestor ExpressionStatement node.
+ */
+function isStartOfExpressionStatement(node) {
+    const start = node.range[0];
+    let ancestor = node;
+
+    while ((ancestor = ancestor.parent) && ancestor.range[0] === start) {
+        if (ancestor.type === "ExpressionStatement") {
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * Determines whether an opening parenthesis `(` or bracket `[` needs to be preceded by a semicolon.
+ * This opening parenthesis or braket should be at the start of an `ExpressionStatement` or at the start of the body of an `ArrowFunctionExpression`.
```

Comment:
Typo braket -> bracket.

While here, maybe we could list all problematic characters: `(`, `[`, `` ` ``, `+`, `-`, `/`.

Reply:
`+`, `-` and `/` don't quite work because they don't trigger ASI after a postfix `++` or `--`. E.g.:

```js
a++ // no ASI here, same as (a++) - 1
-1
```

``` ` ``` seems to fit, so I've added it to the list. Thanks!

Reply:
Correct again. I was looking at the `semi` rule where all 6 are treated the same, but it doesn't special-case preceding postfix operators:

```js
/* eslint semi: ["error", "never", { beforeStatementContinuationChars: "never" }] */

a++; // ok
+b

a++; // ok
-b

a++; // ok
/b/

a++; // looks like false negative
(b)

a++; // looks like false negative
[b]

a++; // looks like false negative
`b`
```